File: airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/book_keeping/SyncStatsBuilder.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.internal.book_keeping;

import io.airbyte.config.StreamSyncStats;
import io.airbyte.config.SyncStats;
import io.airbyte.protocol.models.AirbyteStreamNameNamespacePair;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import lombok.extern.slf4j.Slf4j;

/**
 * Helper function to prepare stats for persistence.
 */
@Slf4j
public class SyncStatsBuilder {

  /**
   * Extract StreamStats from SyncStatsTracker.
   */
  public static List<StreamSyncStats> getPerStreamStats(final SyncStatsTracker syncStatsTracker, final boolean hasReplicationCompleted) {
    // assume every stream with stats is in streamToEmittedRecords map
    return syncStatsTracker.getStreamToEmittedRecords().keySet().stream().map(stream -> {
      final SyncStats syncStats = new SyncStats()
          .withRecordsEmitted(syncStatsTracker.getStreamToEmittedRecords().get(stream))
          .withBytesEmitted(syncStatsTracker.getStreamToEmittedBytes().get(stream))
          .withSourceStateMessagesEmitted(null)
          .withDestinationStateMessagesEmitted(null);

      if (hasReplicationCompleted) {
        syncStats.setBytesCommitted(syncStatsTracker.getStreamToEmittedBytes().get(stream));
        syncStats.setRecordsCommitted(syncStatsTracker.getStreamToEmittedRecords().get(stream));
      } else {
        syncStats.setBytesCommitted(readValueIfPresent(syncStatsTracker.getStreamToCommittedBytes(), stream));
        syncStats.setRecordsCommitted(readValueIfPresent(syncStatsTracker.getStreamToCommittedRecords(), stream));
      }
      return new StreamSyncStats()
          .withStreamName(stream.getName())
          .withStreamNamespace(stream.getNamespace())
          .withStats(syncStats);
    }).collect(Collectors.toList());
  }

  private static Long readValueIfPresent(final Optional<Map<AirbyteStreamNameNamespacePair, Long>> optMap,
                                         final AirbyteStreamNameNamespacePair stream) {
    return optMap.isPresent() ? optMap.get().get(stream) : null;
  }

  /**
   * Extract total stats from SyncStatsTracker.
   */
  public static SyncStats getTotalStats(final SyncStatsTracker syncStatsTracker, final boolean hasReplicationCompleted) {
    final SyncStats totalSyncStats = new SyncStats()
        .withRecordsEmitted(syncStatsTracker.getTotalRecordsEmitted())
        .withBytesEmitted(syncStatsTracker.getTotalBytesEmitted())
        .withSourceStateMessagesEmitted(syncStatsTracker.getTotalSourceStateMessagesEmitted())
        .withDestinationStateMessagesEmitted(syncStatsTracker.getTotalDestinationStateMessagesEmitted())
        .withMaxSecondsBeforeSourceStateMessageEmitted(syncStatsTracker.getMaxSecondsToReceiveSourceStateMessage())
        .withMeanSecondsBeforeSourceStateMessageEmitted(syncStatsTracker.getMeanSecondsToReceiveSourceStateMessage())
        .withMaxSecondsBetweenStateMessageEmittedandCommitted(syncStatsTracker.getMaxSecondsBetweenStateMessageEmittedAndCommitted().orElse(null))
        .withMeanSecondsBetweenStateMessageEmittedandCommitted(syncStatsTracker.getMeanSecondsBetweenStateMessageEmittedAndCommitted().orElse(null));

    if (hasReplicationCompleted) {
      totalSyncStats.setBytesCommitted(totalSyncStats.getBytesEmitted());
      totalSyncStats.setRecordsCommitted(totalSyncStats.getRecordsEmitted());
    } else {
      totalSyncStats.setBytesCommitted(syncStatsTracker.getTotalBytesCommitted().orElse(null));
      totalSyncStats.setRecordsCommitted(syncStatsTracker.getTotalRecordsCommitted().orElse(null));
    }
    return totalSyncStats;
  }

}

File: airbyte-commons-worker/src/main/kotlin/io/airbyte/workers/internal/bookkeeping/SyncStatsBuilder.kt
@file:JvmName("SyncStatsBuilder")

package io.airbyte.workers.internal.bookkeeping

import io.airbyte.config.StreamSyncStats
import io.airbyte.config.SyncStats

/**
 * Extract StreamStats from SyncStatsTracker.
 */
fun SyncStatsTracker.getPerStreamStats(hasReplicationCompleted: Boolean): List<StreamSyncStats> {
  // for readability purposes, avoids having to use `this@getPerStreamStats` in the map below
  val tracker = this

  // assume every stream with stats is in streamToEmittedRecords map
  return getStreamToEmittedRecords().map { (stream, records) ->
    val syncStats: SyncStats =
      SyncStats()
        .withBytesEmitted(tracker.getStreamToEmittedBytes()[stream])
        .withRecordsEmitted(records)
        .withSourceStateMessagesEmitted(null)
        .withDestinationStateMessagesEmitted(null)
        .apply {
          if (hasReplicationCompleted) {
            bytesCommitted = tracker.getStreamToEmittedBytes()[stream]
            recordsCommitted = tracker.getStreamToEmittedRecords()[stream]
          } else {
            bytesCommitted = tracker.getStreamToCommittedBytes()[stream]
            recordsCommitted = tracker.getStreamToCommittedRecords()[stream]
          }
        }

    StreamSyncStats()
      .withStreamName(stream.name)
      .withStreamNamespace(stream.namespace)
      .withStats(syncStats)
  }
    .toList()
}

/**
 * Extract total stats from SyncStatsTracker.
 */
fun SyncStatsTracker.getTotalStats(hasReplicationCompleted: Boolean): SyncStats {
  return SyncStats()
    .withRecordsEmitted(getTotalRecordsEmitted())
    .withBytesEmitted(getTotalBytesEmitted())
    .withSourceStateMessagesEmitted(getTotalSourceStateMessagesEmitted())
    .withDestinationStateMessagesEmitted(getTotalDestinationStateMessagesEmitted())
    .withMaxSecondsBeforeSourceStateMessageEmitted(getMaxSecondsToReceiveSourceStateMessage())
    .withMeanSecondsBeforeSourceStateMessageEmitted(getMeanSecondsToReceiveSourceStateMessage())
    .withMaxSecondsBetweenStateMessageEmittedandCommitted(getMaxSecondsBetweenStateMessageEmittedAndCommitted())
    .withMeanSecondsBetweenStateMessageEmittedandCommitted(getMeanSecondsBetweenStateMessageEmittedAndCommitted())
    .apply {
      if (hasReplicationCompleted) {
        bytesCommitted = bytesEmitted
        recordsCommitted = recordsEmitted
      } else {
        bytesCommitted = getTotalBytesCommitted()
        recordsCommitted = getTotalRecordsCommitted()
      }
    }
}
